### PR TITLE
EUTXO rewrite

### PR DESCRIPTION
**Pull request by polinavino**

### CODE REVIEW

File: plutus-book/doc/09-extended.adoc [hunk 1]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
-tool to make sure a transaction is performing only authorized actions.
-The purpose of witnessing for (non-script) spending of outputs is to verify
+The purpose of witnessing is to verify
 that the wallet spending the output has the private key associated with the address
-in the corresponding UTxO entry (i.e. the address from which it is spending the
+in the corresponding UTXO entry (i.e. the address from which it is spending the
 money) by checking that it was indeed signed (witnessed)
 by that key.
+Scripts, like witnesses, are another
+tool to ensure a transaction is spending only outputs it is authorized to.
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: I think you're conflating witnesses and signatures.

I think *witnesses* are a piece of data attached to a transaction that shows that the transaction is authorized to spend from a particular address, with *signatures* then being the witnesses for public key addresses, and *validators* (plus the redeemer on the input) being the witnesses for script addresses.

Might want to check the ledger spec again to see how they use it.

File: plutus-book/doc/09-extended.adoc [hunk 2]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
+
+A script to which it
+is possible to pay has its own address in the UTXO, and it is the hash of this script.
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: Addresses aren't in the UTXO

File: plutus-book/doc/09-extended.adoc [hunk 3]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
+Only spending from these _script addresses_ can be controlled by the script (whose
+address this is). Let us look at a very small example that omits some of the
+details, but gives the reader an idea of the kind of thing scripts can do.
+Suppose the owner of (non-script) address `a` has promised to pay `x` Ada
+to the first person to guess a word they are thinking of.
+This is a kind of contract, and we can interpret it as a smart contract by
+considering exactly how it controls the transfer of funds. This process
+always starts with a contribution of funds to the contract.
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: Aside, but I really don't find the allusions to a vague separate notion of "smart contract" helps much, and especially here it seems pretty vacuous.

File: plutus-book/doc/09-extended.adoc [hunk 4]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
+locked at the contract address if the guess is correct.
+To express this, we write in `PlutusTx` the following contract,
+representing the condition under which the transfer of funds can take place:
+
+----
+s = "given the hash hs of a secret word, and a word guess wg,
+check that wg hashes to hs"
+----
+
+The contract is created when the owner of `a` pays to this script (i.e. to the
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: What does it mean to say that the "contract is created"? I really think this muddles things.

File: plutus-book/doc/09-extended.adoc [hunk 5]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
+Only spending from these _script addresses_ can be controlled by the script (whose
+address this is). Let us look at a very small example that omits some of the
+details, but gives the reader an idea of the kind of thing scripts can do.
+Suppose the owner of (non-script) address `a` has promised to pay `x` Ada
+to the first person to guess a word they are thinking of.
+This is a kind of contract, and we can interpret it as a smart contract by
+considering exactly how it controls the transfer of funds. This process
+always starts with a contribution of funds to the contract.
+
+Recall that in the classic model, UTXO entries are of the form
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: This seems like just the opposite of what we discussed in our call: rather than talking about the dataflow graph model, we're jumping right in with the formal representation.

File: plutus-book/doc/09-extended.adoc [hunk 6]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
+address `hash s`) `c` Ada. So, he produces a transaction that contains some inputs spending
+UTXO entries with `a` in the address field, and an output
+
+----
+ix |-> (hash s, c, hs)
+----
+
+Note that the hash of the secret word `hs` is a separate piece of data in the
+output - we will explain
+why in more detail when we give the validator-redeemer-data script model overview.
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: I recommend doing an example where the hash is baked into the validator script at first. This allows you to explain validators and redeemers without needing data scripts.

File: plutus-book/doc/09-extended.adoc [hunk 7]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
+word `wg` that hashes to `hs`. So, it can
+be spent by a transaction containing an input which spends the output
+corresponding to `(txid, ix)` and also carries the correct guess `wg`, so the
+input data is,
+
+----
+(txid, ix, wg)
+----
+
+At a node processing this transaction, then, the computation applying
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]:
```suggestion
At a node processing this transaction the computation applying
```

Comment (michaelpj) [reply plutus-book/doc/09-extended.adoc]: I find that, like you, I naturally tend to write with a lot of subordinate clauses, extra asides, things like this (maybe because that's how I speak). But I'm fairly convinced that they're not helpful in writing. So I highly recommend the (painful) process of just going through all your writing and ruthlessly deleting clauses, adverbs etc. that don't add to the sense.

Of course there's a balance to be struck here with having some writing style, but my experience is that it's much easier to err on the side of too much than too little.

Somewhat exaggerated, but https://youtu.be/8TD-20Mb_7M?t=3161 is a useful example of editing.

File: plutus-book/doc/09-extended.adoc [hunk 8]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
-
-The wallet is expected to compute the fees and either add them to the inputs, or
-subtract them to the outputs if possible. The exact behaviour should be
-configurable by the user. Note that changing the inputs and outputs affects the
-transaction fees. Note also, however, that the fee can always be pre-computed
+Let us not forget transaction fees here - the wallet is expected to compute the
+fees and either add them to the inputs, or
+subtract them to the outputs if possible.
+Note that even with the addition of scripts into transaction data,
+the fee can always be pre-computed
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: I think https://github.com/input-output-hk/plutus/pull/1379/files#diff-812cebcb9ddc363daff7545c3d719831R121 still applies.

File: plutus-book/doc/09-extended.adoc [hunk 9]
```diff
@@ -62,83 +66,133 @@ sections of a Haskell program, called Plutus Tx, which are then compiled into
...
-before the application. Let us now look at the function of validator
-scripts and their arguments on the blockchain.
+arguments (which are also scripts). Let us now look at the function of validator
+scripts and their arguments on the blockchain, then we will explain how our
+small example in the previous section fits into this model.
 
 . Validator script
 * Carried by a transaction spending funds from a script address
-* Not stored in full in the ledger UTxO
+* Defines the _logic_ or _rules_ of a contract
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: What does this mean?

File: plutus-book/doc/09-extended.adoc [hunk 10]
```diff
@@ -177,114 +225,64 @@ together out of transaction and ledger data,
...
-This is how the outputs are structured both inside the carrying transaction
-and the UTxO. In the extended model, the position of an output in the outputs list
-carried by a transaction is its reference index `ix`.
-In the UTxO, the outputs are stored as a finite
-map with the keys of the form `(txid, ix)`. Once a transaction is validated,
-the update mechanism
-of the UTxO is the same for both the the extended and classic model (with the
-extended model storing the additional `TxOutType` data described above).
-For a reminder, see the <<03-UTXO#03-UTXO>> chapter.
+evaluates to `True`. The `PendingTx` value is not used in this example, but
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: The `PendingTx` value isn't mentioned.

File: plutus-book/doc/09-extended.adoc [hunk 11]
```diff
@@ -177,114 +225,64 @@ together out of transaction and ledger data,
...
+we could have extended the contract logic to only allow a transaction to
+spend the funds under the additional condition that it is doing so after
+after a certain slot. This `currentSlot` data would be passed via `PendingTx`.
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: You need to say a little more for this to be comprehensible to someone who doesn't know what's in `PendingTx`. Plus this makes it sound as though `PendingTx` contains the current slot, which it doesn't.

File: plutus-book/doc/09-extended.adoc [hunk 12]
```diff
@@ -177,114 +225,64 @@ together out of transaction and ledger data,
...
+The address of the funds locked by the contract is always the _validator script
+hash_. In this case, we could have included the hashed secret word directly
+in the validator. Making it part of the data script (state) of the contract
+makes the same validator usable for multiple word games (this is not the case
+for most other more complex contracts). The use of the
+redeemer, however, is inevitable here. It has to come from a participant,
+and represents the act of guessing the secret word.
 
-To recap the paying and spending process in a different way,
+We just saw the the extension to the UTxO and transaction data necessary
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: Should this be in another section?

File: plutus-book/doc/09-extended.adoc [hunk 13]
```diff
@@ -177,114 +225,64 @@ together out of transaction and ledger data,
...
 `i` references
 - we have the validator, redeemer and data script all available
-- if it succeeds, the transaction is allowed to spend the UTxO output `i` references
+- if it succeeds, the transaction is allowed to spend the UTXO output `i` references
 - validation is performed on-chain
 
-If the `runScript` computation returns true for a specific output being spent,
-the validation process proceeds.
+If the validation computation returns true for a specific script-address UTXO
+entry, the validation process proceeds.
```
Comment (michaelpj) [inline plutus-book/doc/09-extended.adoc]: Validation doesn't apply to UTXO entries. Perhaps you mean for the particular pair of an input and an output?

File: plutus-book/doc/09-extended.adoc (hunk 3 above)
Comment (michaelpj) [reply plutus-book/doc/09-extended.adoc]: In fact, I think it's actively confusing every time it appears in this chapter.

File: plutus-book/doc/09-extended.adoc (hunk 5 above)
Comment (michaelpj) [reply plutus-book/doc/09-extended.adoc]: Have a look at Manuel's Lambda Jam talk for another approach to introducing validators without getting bogged down in details: https://youtu.be/96o8G5dVCaM?t=52

**Comment by michaelpj** [comment]
Perhaps a good thing to write would be a clear outline with all the concepts and where they're going to be introduced.

File: marlowe/src/Language/Marlowe/Client.lhs [hunk 14]
```diff
@@ -116,15 +116,24 @@ createContract validator contract value = do
 
 ----
 
-Prepare 'TxIn' and 'TxOut' generator for Marlowe style wallet actions.
-Explain more. A transaction that builds the set of TxIns and TxOuts.
+This function builds the `TxIn` and `TxOut` values for a transaction which
```
[the remark this answers is not in the document]
Comment (polinavino) [reply marlowe/src/Language/Marlowe/Client.lhs]: yup unrelated. I will work on that more when this is ship shape

File: plutus-book/doc/09-extended.adoc [hunk 15]
```diff
@@ -1,33 +1,37 @@
 [#09-extended]
-= Extended UTxO Model
+= Extended UTXO Model
```
[the remark this answers is not in the document]
Comment (polinavino) [reply plutus-book/doc/09-extended.adoc]: I just incorporated them! Thank you!!

File: plutus-book/doc/09-extended.adoc (hunk 5 above)
Comment (polinavino) [reply plutus-book/doc/09-extended.adoc]: I will discuss the UTXO as a dataflow graph model in the UTXO chapter, which I will upload soon. I think it makes more sense there, and go straight to the formal representation in this chapter. Would that make sense?

Comment (michaelpj) [reply plutus-book/doc/09-extended.adoc]: Well, the EUTXO model also has a nice interpretation on the more abstract level. You can talk about all the conceptual stuff about locking outputs, data, having access to the pending transaction etc without getting into the formal stuff at all.

I'm still unconvinced that the formal presentation actually adds much...
